src/Features/CSharp/Portable/UseRecursivePatterns/AnalyzedNode.cs

```diff
@@ -0,0 +1,545 @@
+            }
+        }
+
+        internal sealed class Variable : Evaluation
+        {
+            public readonly ISymbol DeclaredSymbol;
+
+            public Variable(Evaluation? input, ISymbol symbol, SyntaxNode? syntax = null) : base(input, syntax)
+            {
+                Debug.Assert(symbol is ILocalSymbol || symbol is IFieldSymbol);
+                DeclaredSymbol = symbol;
+            }
+
+            protected override Test WithInputCore(Evaluation? newInput)
+            {
+                return new Variable(newInput, DeclaredSymbol, Syntax);
+            }
+        }
+
+        internal sealed class Type : Evaluation
+        {
+            public readonly ITypeSymbol TypeSymbol;
+
+            public Type(Evaluation? input, ITypeSymbol type) : base(input)
+            {
+                TypeSymbol = type;
+            }
+
+            protected override Test WithInputCore(Evaluation? newInput)
+            {
+                return new Type(newInput, TypeSymbol);
+            }
+        }
+
+        internal sealed class IndexEvaluation : Evaluation
+        {
+            public readonly IPropertySymbol Property;
+            public readonly int Index;
+
+            public IndexEvaluation(Evaluation? input, IPropertySymbol property, int index, SyntaxNode? syntax = null)
+                : base(input, syntax)
+            {
+                Property = property;
+                Index = index;
+            }
+
+            protected override Test WithInputCore(Evaluation? newInput)
+            {
+                return new IndexEvaluation(newInput, Property, Index, Syntax);
+            }
+        }
+
+        internal sealed class MemberEvaluation : Evaluation
+        {
+            public readonly ISymbol Symbol;
+
+            private MemberEvaluation(Evaluation? input, ISymbol symbol, SyntaxNode? syntax)
+                : base(input, syntax)
+            {
+                Symbol = symbol;
+            }
+
+            public MemberEvaluation(Evaluation? input, IFieldReferenceOperation op)
+                : this(input, op.Field.CorrespondingTupleField ?? op.Field, op.Syntax)
+            {
+            }
+
+            public MemberEvaluation(Evaluation? input, IPropertyReferenceOperation op)
+                : this(input, op.Property, op.Syntax)
+            {
+            }
+
+            public MemberEvaluation(Evaluation? input, IFieldSymbol field)
+                : this(input, field, syntax: null)
+            {
+                Debug.Assert(field.IsTupleField());
+            }
+
+            public bool IsTupleItemField => Symbol.IsTupleField();
+
+            public bool IsITupleLengthProperty
+            {
+                get
+                {
+                    return Symbol is
+                    {
+                        Kind: SymbolKind.Property,
+                        Name: WellKnownMemberNames.LengthPropertyName,
+                        ContainingType: { Name: "ITuple" }
```

Comment:
This is probably not safe. We should use GetTypeByMetadataName or check the full name (+namespace).
To use metadata type we need a Compilation which is only available in construction time. Need to see how things works out.